src/methods.rs

```diff
@@ -30,16 +33,37 @@ declare_lint!(pub WRONG_SELF_CONVENTION, Warn,
 declare_lint!(pub WRONG_PUB_SELF_CONVENTION, Allow,
               "defining a public method named with an established prefix (like \"into_\") that takes \
                `self` with the wrong convention");
+declare_lint!(pub CALLING_MAIN, Warn,
+              "calling into main detected");
 
 impl LintPass for MethodsPass {
     fn get_lints(&self) -> LintArray {
         lint_array!(OPTION_UNWRAP_USED, RESULT_UNWRAP_USED, STR_TO_STRING, STRING_TO_STRING,
-                    SHOULD_IMPLEMENT_TRAIT, WRONG_SELF_CONVENTION)
+                    SHOULD_IMPLEMENT_TRAIT, WRONG_SELF_CONVENTION, CALLING_MAIN)
     }
 }
 
 impl LateLintPass for MethodsPass {
     fn check_expr(&mut self, cx: &LateContext, expr: &Expr) {
+        fn no_std_present(cx: &LateContext) -> bool {
+        	let attrs : &[Attribute] = &cx.krate.attrs; // TODO replace these two...
+        	for attr in attrs { // lines with ...
+        	// for &attr in cx.krate.attrs { ... once type inf on nightly fixed
+                if let MetaWord(ref mword) = attr.node.value.node {
+                    if mword == &"no_std" { return true; }
+                }
+            }
+            false
+        }
+        if let ExprCall(ref func, ref args) = expr.node {
+            if args.is_empty() {
+                if let ExprPath(_, path) = func.node.clone() {
+                    if !path.global && path.segments[0].identifier.name.as_str() == "main" && !no_std_present(cx) {
```

Comment:
This isn't exactly correct

In the following code, the `main()` call contains an AST path which isn't global.


```rust
fn main() {

}
 mod foo {
  fn main(){}
  fn foo(){
   main();
  }
}
```

A global AST path is one with starts with a double colon, i.e. `::std::foo::bar`.

We need to check the path resolution here. Try using [match_def_path](https://github.com/Manishearth/rust-clippy/blob/master/src/utils.rs#L68) (use `cx.tcx.def_map.get(expr.node.id)` to get the defid -- it should be a `DefFn`, see http://manishearth.github.io/rust-internals-docs/rustc/middle/def/struct.PathResolution.html and the `base_def` field). Also ensure that the DefId [is local](http://manishearth.github.io/rust-internals-docs/rustc/middle/def_id/struct.DefId.html#method.is_local).

The resulting path should either be `crate_name::main` or just plain `main`, I forgot which. Try both, and add an extra test to ensure that nearby non-toplevel functions named `main` don't get caught.